src/business_logic/transaction/objects/internal_invoke_function.rs

```diff
@@ -34,7 +38,52 @@ pub(crate) struct InternalInvokeFunction {
 }
 
 impl InternalInvokeFunction {
-    #[allow(dead_code)]
+    pub fn new(
+        contract_address: Address,
+        entry_point_selector: Felt,
+        max_fee: u64,
+        calldata: Vec<Felt>,
+        signature: Vec<Felt>,
+        chain_id: Felt,
+        nonce: Option<Felt>,
+    ) -> Result<Self, TransactionError> {
+        let version = TRANSACTION_VERSION;
+        let (entry_point_selector_field, additional_data) = preprocess_invoke_function_fields(
+            entry_point_selector.clone(),
+            nonce.clone(),
+            version,
+        )?;
+
+        let hash_value = calculate_transaction_hash_common(
+            TransactionHashPrefix::Invoke,
+            version,
+            contract_address.clone(),
+            entry_point_selector_field.into(),
+            &calldata,
+            max_fee,
+            chain_id,
+            &additional_data,
+        )?;
+
+        let validate_entry_point_selector = felt_str!(
+            "626969833899987279399947180575486623810258720106406659648356883742278317941"
+        );
+
+        Ok(InternalInvokeFunction {
+            contract_address,
+            entry_point_selector,
+            entry_point_type: EntryPointType::External,
+            calldata,
+            tx_type: TransactionType::InvokeFunction,
+            version,
+            max_fee,
+            signature,
+            validate_entry_point_selector,
+            nonce: nonce.unwrap(),
```

Comment:
as far as i understand, it may not have a nonce, but many places assumes that there is one. i am going to make an unwrap_or() and default to a 0 if there is nothing but we should check.

Reply:
Check what Python does before doing anything since we could be introducing bugs without intending to.
Still, if we're going to assume it has a value I'd rather assume it as soon as possible, so this suggestion still applies.

Reply:
python is a bit strange because in some functions it assumes that has type int whether others are called as Optional. i will check it again anyway.